Publish adjustment algorithms for ISIS SANS

AntonPiccardoSelg opened the pull request:
Fixes #19340

**Important**
The changes here rely on #19333 , do not review before this has been merged

This PR exposes the algorithms relevant for the creation of the adjustment workspaces.

**To test:**
Please code review. These changes are part of larger set of complete changes which are currently being tested by the ISIS SANS team.

*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

peterfpeterson commented:
Didn't we talk about large PRs being hard to test on [2017-04-04](https://github.com/mantidproject/documents/blob/master/Project-Management/MantidReviewMeeting/2017/review-meeting-2017-4-4.md)? My opinion is that 5.6k LOC is big.

AntonPiccardoSelg commented:
@peterfpeterson  Yes, i agree. This PR is following your suggestion as presented in the MRM. The PR is preceded by PR #19333. At the beginning of this PR it says that the PR should not be reviewed until #19333 has been merged. This effectively halves the LOC of this PR.

AntonPiccardoSelg commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANS/SANSCreateAdjustmentWorkspaces.py:
```diff
@@ -0,0 +1,228 @@
+        if wave_length_and_pixel_adjustment_workspace:
+            self.setProperty("OutputWorkspaceWavelengthAndPixelAdjustment", wave_length_and_pixel_adjustment_workspace)
+
+        # TODO set diagnostic output workspaces
```
Yes, there are several TODOs in there which refer to "nice-to-have" features which we can add once we have freed up resources. I will expand the TODOs

AntonPiccardoSelg commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANS/SANSNormalizeToMonitor.py:
```diff
@@ -0,0 +1,216 @@
+
+        convert_alg = create_unmanaged_algorithm(convert_name, **convert_options)
+        convert_alg.setPropertyValue("OutputWorkspace", EMPTY_NAME)
+        convert_alg.setProperty("OutputWorkspace", workspace)
```
That is a good question.

To answer your second part first:  `convert_options` is a `dict` which gets passed into `create_unmanaged_algorithm` which then take the keys of the `dict` as the property name and the value of the `dict` as the property value and sets it on `alg.setProperty`. In the case above I need to set a string (see reason below) using the `alg.setPropertyValue`. I could have added have added `{"OutputWorkspace": workspace}` to the `convert_options, but since we have to set `OutputWorkspace` twice here, I wanted to leave it together.

Regarding your first question, the answer is slightly complex (but good to know). When Antii reviewed my last PR he had the same question. My reply was:

>The algorithms are run in general with unnamed workspaces, ie workspaces which are not placed on the ADS. You might have seen that the `OutputWorkspace` property needs to be set, something like
```python
...
alg.setProperty("InputWorkspace", ws)
...
alg.setProperty("OutputWorkspace", "test")
alg.execute()
out_ws = alg.getProperty("OutputWorkspace").value
....
```
>Mantid needs names for the `OutputWorkspace` property even if they are only temporary. It also means that output workspace is most likely a copy of the input. In my case here I want to perform an in-place operation on an unnamed workspace, hence I need to provide the workspace itself to the output and I need to provide a temporary name.

>The example below might illustrate this
```python
# -----------------------------------------------------------------------------------------------------------------------
# Create a fake algorithm which peforms some in-place operation
# -----------------------------------------------------------------------------------------------------------------------
class FakeAlgorithm(DataProcessorAlgorithm):
    def PyInit(self):
        self.declareProperty(MatrixWorkspaceProperty("InputWorkspace", '', optional=PropertyMode.Mandatory, direction=Direction.Input), doc='input')
    def PyExec(self):
        ws = self.getProperty("InputWorkspace").value
        
        alg = AlgorithmManager.createUnmanaged("ConvertUnits")
        alg.initialize()
        alg.setChild(True)
        alg.setProperty("InputWorkspace", ws)
        alg.setProperty("Target", "Wavelength")
        # UNCOMMENT LINE BELOW TO MAKE IT WORK
        #alg.setPropertyValue("OutputWorkspace", "EMPTY")
        alg.setProperty("OutputWorkspace", ws)
        
        alg.execute()
        ws_converted = alg.getProperty("OutputWorkspace").value

AlgorithmFactory.subscribe(FakeAlgorithm)


# -----------------------------------------------------------------------------------------------------------------------
# Provide an unnamed workspace
# -----------------------------------------------------------------------------------------------------------------------
alg_created = AlgorithmManager.createUnmanaged("CreateSampleWorkspace")
alg_created.initialize()
alg_created.setChild(True)
alg_created.setProperty("OutputWorkspace", "dummy")
alg_created.execute()
out_ws = alg_created.getProperty("OutputWorkspace").value


# -----------------------------------------------------------------------------------------------------------------------
# Run the algorithm
# -----------------------------------------------------------------------------------------------------------------------
alg_fake = AlgorithmManager.createUnmanaged("FakeAlgorithm")
alg_fake.initialize()
alg_fake.setChild(True)
alg_fake.setProperty("InputWorkspace", out_ws)
alg_fake.execute()
```
>You should see that if you uncomment the marked line, then the algorithm runs without an issue, else it will complain that there has been no name set for the output

gemmaguest replied:
Thanks for the detailed explanation, that's good to know 👍

AntonPiccardoSelg commented on scripts/SANS/sans/algorithm_detail/calculate_transmission_helper.py:
```diff
@@ -0,0 +1,268 @@
+                        "SkipMonitors": True}
+        flat_alg = create_unmanaged_algorithm(flat_name, **flat_options)
+        flat_alg.setPropertyValue("OutputWorkspace", EMPTY_NAME)
+        flat_alg.setProperty("OutputWorkspace", workspace)
```
See above

AntonPiccardoSelg commented:
* Made the `param` format consistent
* Fixed typos
* Fixed logical mistake in test
* Added more detailed explanation to TODOs in code
* Updated documenation.

gemmaguest commented:
This all looks good now, please :shipit:
